### DISCUSSION TITLE
Why `get_cog` is annotated to return `Optional[Cog]`?

### DISCUSSION THREAD

**Discussion by Aman-14** [Rapptz/discord.py#7713] [Q&A]
Get cog is annotated to return `Optional[Cog]` but it returns a subclass of `Cog`.
![image](https://user-images.githubusercontent.com/72304680/158962042-014f5a63-a1a7-40ca-97eb-241a00818593.png)

Adding `get_cog` in my bot class with bound return type solved this problem, just wanted to know why its annotated that way.
![image](https://user-images.githubusercontent.com/72304680/158962366-909d7ef5-be5e-44e4-bdf9-b30af287ad94.png)

**Comment by Rapptz** [comment]
The internal mapping is stored as `Dict[str, Cog]`. Therefore `get_cog` returns `Optional[Cog]`. Your suggestion of doing `Optional[CogT]` would mean that the mapping would have to be `Dict[str, CogT]` and this generic would have to be propagated to the `Bot` class as a generic. However, it would also mean that rather than just using polymorphism it also uses monomorphism which would make *all* cogs be a subtype of whatever `CogT` generic is passed to the bot. When `CogT` is `Cog` this is more or less the same thing.

Likewise, if `Reminder` is a subtype of `Cog` then you can just downcast it. I suggest getting a different type checker like Pyright which understands code like this.
